### PR TITLE
[MRG] Allow electrodes.tsv to not fully match Raw object

### PR DESCRIPTION
**Pull request by adam2392**
PR Description
--------------

Closes: #701 
~~Partially addresses: #605~~
Similar in style to #698 

# Summary
This PR allows electrodes.tsv contents to not fully match chs in Raw. 

In addition, it fixes the repo to match recent clarifications to the spec and validator. Since the [spec](https://bids-specification.readthedocs.io/en/latest/04-modality-specific-files/04-intracranial-electroencephalography.html#electrode-description-_electrodestsv) and validator was updated to constrain how `space` entity and `*CoordinateSystem` in the `*coordsystem.json` file must match, there are a few changes needed to make things work.

1. **Separation of "BIDS" coordinate frames and "MNE" coordinate frames:** unfortunately, there just isn't a 1-1 matching, so we just do the best we can to match w/e coordinate frames possible. Note there is a loss going back and forth (e.g. captrak/electkneuromag all == 'head' in mne)
2. **Casing:** BIDS coordinate system names are case-sensitive in spec and validator. We should therefore get rid of any "lower-casing", since strings must match exactly. A bit stringent... but it's in spec/validator.

## iEEG coordinate frames:
1. iEEG coordinate frames only support `mni_tal` for 'fsaverage'` and `unknown` for  `"Other"/<standard template identifiers checked by validator>`, and `ras` for `ACPC`. Note that `ACPC` in BIDS maps to `ras` in MNE, but not the other way around.
2. `fs_tal` in mne-python is unused, per discourse discussion.

## Unit tests
- Some of the read/write unit tests dealing with `space` entity and coordinatesystem writing needed to be fixed.

Old Note (not implemented, so no need to read)
------------------------------------------------
If we don't think having `read_dig_bids` as a public function is necessary, then I can take it out, but the general workflow in #605 related to iEEG electrode coordinate processing doesn't necessitate a Raw file. In addition, if mne-python coordinate system handling becomes more and more in-line with BIDS, then the DigMontage object will be valuable. For example, I'm imagining a future where:

```
# all workflows not requiring the Raw file itself yet
montage_mm = read_dig_bids()

# convert to voxel space
montage_vox = montage_mm.convert_to('voxel')

# maybe convert to tkras space
montage_mm = montage_mm.convert_to('tkras')

# maybe utilize the IntendedFor parameter to get the filepath to the Nifti image to interpret iEEG coords on
intended_img_fpath = montage_mm.get_intended_for()
t1w_img = nb.load(intended_img_fpath)
```

In summary, having an API to go from BIDS files -> DigMontage is a nice interface for people that then need to maybe do other stuff that doesn't need a Raw dataset.

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [ ] All comments resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [ ] PR title starts with [MRG]
- [ ] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/master/doc/whats_new.rst) is updated
- [ ] PR description includes phrase "closes <#issue-number>"

### CODE REVIEW

**Comment by adam2392** [comment]
@hoechenberger I have taken out the explicit checks for subset of electrodes in raw, per #485

**Comment by hoechenberger** [comment]
Thanks @adam2392, I will take a closer look at this tomorrow!

**Review by sappelhoff — COMMENTED** [review]
@Immiora can you please try if the changes in this PR solve your problem?

File: mne_bids/dig.py [hunk 1]
```diff
@@ -25,7 +25,7 @@
 
 
 def _handle_electrodes_reading(electrodes_fname, coord_frame,
-                               coord_unit, raw, verbose):
+                               coord_unit, verbose):
     """Read associated electrodes.tsv and populate raw.
```
Comment (sappelhoff) [inline mne_bids/dig.py:29]: you are not populating "raw" anymore, right?

File: mne_bids/dig.py [hunk 2]
```diff
@@ -34,30 +34,8 @@ def _handle_electrodes_reading(electrodes_fname, coord_frame,
...
-               )
-
-        # XXX: this could be due to MNE inserting a 'STI 014' channel as the
-        # last channel: In that case, we can work. --> Can be removed soon,
-        # because MNE will stop the synthesis of stim channels in the near
-        # future
-        if not (ch_names_raw[-1] == 'STI 014' and
-                ch_names_raw[:-1] == ch_names_tsv):
-            raise RuntimeError(msg)
-
```
Comment (sappelhoff) [inline mne_bids/dig.py:60]: why did you remove this? did we have an issue with that somewhere?

I mean, it's RECOMMENDED in BIDS ... but we could still raise a warning if someone doesn't comply?

File: doc/whats_new.rst [hunk 3]
```diff
@@ -35,6 +35,8 @@ Enhancements
...
 - When reading data where the same event name or trial type refers to different event or trigger values, we will now create a hierarchical event name in the form of ``trial_type/value``, e.g. ``stimulus/110``  by `Richard Höchenberger`_ (:gh:`688`)
 - When reading data via :func:`mne_bids.read_raw_bids`, the channel names specified in the BIDS ``*_channels.tsv`` files now always take precedence (and do not need to match) the channel names stored in the raw files anymore, by `Richard Höchenberger`_ (:gh:`691`)
+- When reading data via :func:`mne_bids.read_raw_bids`, the channel names specified in the BIDS ``*_electrodes.tsv`` files now always take precedence (and do not need to match) the channel names stored in the raw files anymore, by `Adam Li`_ (:gh:`704`)
+- Add :func:`mne_bids.read_dig_bids` to allow reading of a ``DigMontage`` object directly from MEG/EEG/iEEG BIDS ``*electrodes.tsv`` and ``*coordsystem.json`` files, by `Adam Li`_ (:gh:`704`)
```
Comment (sappelhoff) [inline doc/whats_new.rst:39]: just a note: I introduced a typo in https://github.com/mne-tools/mne-bids/commit/45f3f39c54472b3765692ef1fdfd285f1b6172f2, (missing a `)` in the changelog entry), please add that once you rebase here :) :) :)

**Review by sappelhoff — COMMENTED** [review]
The changes overall seem fine to me, but I think that making a `read_dig_bids` function public we are catering to a very niche use case that can also be solved with two or three lines of code by the niche-user (adam :-P)

I am not 100% against it, but more public functions are not always better.

File: doc/whats_new.rst (hunk 3 above)
Comment (hoechenberger) [reply doc/whats_new.rst:39]: I also think lines 37 and 38 should be merged into a single entry

File: mne_bids/dig.py (hunk 2 above)
Comment (hoechenberger) [reply mne_bids/dig.py:60]: I removed a similar test in #698, so if we change this one here to a warning, we should revert the relevant part to #698 and change it into a warning too

File: doc/whats_new.rst (hunk 3 above)
Comment (sappelhoff) [reply doc/whats_new.rst:39]: and @eort requests an edit to .mailmap: https://github.com/mne-tools/mne-bids/commit/45f3f39c54472b3765692ef1fdfd285f1b6172f2#r47542622

File: mne_bids/dig.py (hunk 1 above)
Comment (adam2392) [reply mne_bids/dig.py:29]: Yes, I think since `_read_dig_bids` is the parent function to coordinate the `*coordsystem.json` and `*electrodes.tsv`, then the setting of raw should occur outside.

**Comment by adam2392** [comment]
> The changes overall seem fine to me, but I think that making a `read_dig_bids` function public we are catering to a very niche use case that can also be solved with two or three lines of code by the niche-user (adam :-P)
> 
> I am not 100% against it, but more public functions are not always better.

I removed it

File: doc/whats_new.rst (hunk 3 above)
Comment (adam2392) [reply doc/whats_new.rst]: Done and done. lmk if it looks good.

File: mne_bids/dig.py (hunk 2 above)
Comment (adam2392) [reply mne_bids/dig.py:60]: I'm okay with no warning here if no warning in #698 . 

Lmk what you guys decide?

File: doc/whats_new.rst [hunk 4]
```diff
@@ -58,7 +58,8 @@ Bug fixes
 - :func:`mne_bids.read_raw_bids` now ignores ``exclude`` parameters passed via ``extra_params``, by `Richard Höchenberger`_ (:gh:`703`)
 - :func:`mne_bids.write_raw_bids` now retains original event IDs in the ``value`` column of ``*_events.tsv``, by `Richard Höchenberger`_ (:gh:`708`)
 - Fix writing correct ``iEEGCoordinateSystemDescription``, by `Stefan Appelhoff`_ (:gh:`706`)
-- FIF files that were split due to filesize limitations (using the ``_split-<label>`` entity, are now all listed in ``scans.tsv``, as recommended by BIDS, by `Eduard Ort`_ (:gh:`710`)
+- FIF files that were split due to filesize limitations (using the ``_split-<label>`` entity, are now all listed in ``scans.tsv``, as recommended by BIDS), by `Eduard Ort`_ (:gh:`710`)
+
```
Comment (sappelhoff) [inline doc/whats_new.rst:62]:
```suggestion
- FIF files that were split due to filesize limitations (using the ``_split-<label>`` entity), are now all listed in ``scans.tsv``, as recommended by BIDS, by `Eduard Ort`_ (:gh:`710`)

```

File: mne_bids/dig.py [hunk 5]
```diff
@@ -527,10 +497,26 @@ def _read_dig_bids(electrodes_fpath, coordsystem_fpath,
...
+        nan_chs = []
+        for ch_name, ch_coord in ch_pos.items():
+            if any(np.isnan(ch_coord)) and ch_name not in raw.info['bads']:
+                nan_chs.append(ch_name)
+        if len(nan_chs) > 0:
+            warn("There are channels without locations "
+                 "(n/a) that are not marked as bad: {}".format(nan_chs))
+
+    # add montage to Raw object
+    raw.set_montage(montage, on_missing='warn', verbose=verbose)
```
Comment (sappelhoff) [inline mne_bids/dig.py:520]: either you need to add an "if clause" (if raw is available ...), or you need to mention in the docstr that raw is REQUIRED ... currently it says OPTIONAL

File: mne_bids/dig.py [hunk 6]
```diff
@@ -451,18 +420,19 @@ def _read_dig_bids(electrodes_fpath, coordsystem_fpath,
...
     coordsystem_fpath : str
         Filepath of the coordsystem.json to read.
-    raw : mne.io.Raw
-        The data as MNE-Python Raw object.
     datatype : str
         Type of the data recording. Can be ``meg``, ``eeg``,
         or ``ieeg``.
+    raw : mne.io.Raw | None
+        Optional. If passed in, should be the raw data as MNE-Python
+        Raw object. Will set montage read in via ``raw.set_montage(montage)``.
```
Comment (sappelhoff) [inline mne_bids/dig.py:428]: see comment below, I think this can be made REQUIRED for now.

File: mne_bids/dig.py (hunk 2 above)
Comment (sappelhoff) [reply mne_bids/dig.py:60]: okay, fine to remove it then

File: mne_bids/dig.py [hunk 7]
```diff
@@ -150,7 +118,7 @@ def _get_impedances(raw, names):
     return impedances
 
 
-def _electrodes_tsv(raw, fname, datatype, overwrite=False, verbose=True):
+def _write_electrodes_tsv(raw, fname, datatype, overwrite=False, verbose=True):
```
Comment (sappelhoff) [inline mne_bids/dig.py:121]: why the renaming?

File: mne_bids/dig.py [hunk 8]
```diff
@@ -226,8 +194,9 @@ def _electrodes_tsv(raw, fname, datatype, overwrite=False, verbose=True):
     _write_tsv(fname, data, overwrite=True, verbose=verbose)
 
 
-def _coordsystem_json(*, raw, unit, hpi_coord_system, sensor_coord_system,
-                      fname, datatype, overwrite=False, verbose=True):
+def _write_coordsystem_json(*, raw, unit, hpi_coord_system,
+                            sensor_coord_system, fname, datatype,
+                            overwrite=False, verbose=True):
```
Comment (sappelhoff) [inline mne_bids/dig.py:199]: why the renaming?

**Review by sappelhoff — COMMENTED** [review]
thanks for removing the controversial part :)

File: mne_bids/dig.py (hunk 7 above)
Comment (hoechenberger) [reply mne_bids/dig.py:121]: Personally I think the new name makes things clearer, but I'd also vote for reverting this change here and make a new PR if this is really important to you

File: doc/whats_new.rst [hunk 9]
```diff
@@ -34,7 +34,7 @@ Enhancements
 
 - Some datasets out in the real world have a non-standard ``stim_type`` instead of a ``trial_type`` column in ``*_events.tsv``. :func:`mne_bids.read_raw_bids` now makes use of this column, and emits a warning, encouraging users to rename it, by `Richard Höchenberger`_ (:gh:`680`)
 - When reading data where the same event name or trial type refers to different event or trigger values, we will now create a hierarchical event name in the form of ``trial_type/value``, e.g. ``stimulus/110``  by `Richard Höchenberger`_ (:gh:`688`)
-- When reading data via :func:`mne_bids.read_raw_bids`, the channel names specified in the BIDS ``*_channels.tsv`` files now always take precedence (and do not need to match) the channel names stored in the raw files anymore, by `Richard Höchenberger`_ (:gh:`691`)
+- When reading data via :func:`mne_bids.read_raw_bids`, the channel names specified in the BIDS ``*_channels.tsv`` and ``*_electrodes.tsv`` files now always take precedence (and do not need to match) the channel names stored in the raw files anymore, by `Richard Höchenberger`_ (:gh:`691`) and by `Adam Li`_ (:gh:`704`)
```
Comment (hoechenberger) [inline doc/whats_new.rst:37]:
```suggestion
- When reading data via :func:`mne_bids.read_raw_bids`, the channel names specified in the BIDS ``*_channels.tsv`` and ``*_electrodes.tsv`` files now always take precedence (and do not need to match) the channel names stored in the raw files anymore, by `Adam Li`_ and `Richard Höchenberger`_ (:gh:`691`, :gh:`704`)
```

File: mne_bids/dig.py (hunk 6 above)
Comment (hoechenberger) [reply mne_bids/dig.py:428]: It's not optional anyway, as it doesn't have a default value in the signature.

File: mne_bids/dig.py (hunk 7 above)
Comment (adam2392) [reply mne_bids/dig.py:121]: I mainly renamed because it was more clear for dev purposes.

File: mne_bids/config.py [hunk 10]
```diff
@@ -172,21 +172,26 @@
...
-    'voxel': 'mri_voxels',
-    'mri': 'mri',
-    'unknown': 'unknown'
+    'CTF': 'ctf_head',
+    '4DBti': 'ctf_head',
+    'KitYokogawa': 'ctf_head',
+    'ElektaNeuromag': 'head',
+    'ChietiItab': 'head',
+    'CapTrak': 'head',
+    'ACPC': 'ras',  # XXX: there is no ACPC in mne-python
```
Comment (hoechenberger) [inline mne_bids/config.py:181]: should this better become `'unknow'` then?

File: mne_bids/config.py [hunk 11]
```diff
@@ -172,21 +172,26 @@
...
+    'fsaverage': 'mni_tal',
+    'MNI305': 'mni_tal',
+    'Other': 'unknown'
+}
+MNE_TO_BIDS_FRAMES = {
+    'ctf_head': 'CTF',
+    'head': 'CapTrak',
+    'mni_tal': 'fsaverage',
+    'fs_tal': 'fsaverage',  # XXX: not used
```
Comment (hoechenberger) [inline mne_bids/config.py:190]: What does "not used" mean here?

File: mne_bids/dig.py (hunk 7 above)
Comment (hoechenberger) [reply mne_bids/dig.py:121]: I don't feel strongly against it, so if it helps you, I'm okay with leaving the new names.
@sappelhoff, your call :)

File: mne_bids/config.py (hunk 10 above)
Comment (adam2392) [reply mne_bids/config.py:181]: isn't ACPC assumed to be in RAS? Would it be better to try to get the data annotated as much as possible?

File: mne_bids/config.py (hunk 11 above)
Comment (adam2392) [reply mne_bids/config.py:190]: @larsoner mentioned that there's no reason to use fs_Tal but it's still in mne anyways so perhaps have it map to fsaverage since that's prolly what the user meant?

Let me know if just take it out tho.

File: mne_bids/dig.py (hunk 7 above)
Comment (sappelhoff) [reply mne_bids/dig.py:121]: fine with me as well to leave it in the slightly-more-verbose-but-easier-to-understand version 👍

File: mne_bids/config.py (hunk 10 above)
Comment (hoechenberger) [reply mne_bids/config.py:181]: Ok, I trust your judgment there :)

File: mne_bids/config.py (hunk 11 above)
Comment (hoechenberger) [reply mne_bids/config.py:190]: @larsoner said:

> I wouldn’t use fs_tal, I don’t think you need it.

My suggestion: comment out the `fa_tal` line so we keep a reference, but don't "enable" it until maybe one day we find out we need it.

WDYT?

```suggestion
    # 'fs_tal': 'fsaverage',  # XXX: not used
```

**Comment by hoechenberger** [comment]
@adam2392 Tests are failing. Aside from this, LGTM

**Comment by adam2392** [comment]
> @adam2392 Tests are failing. Aside from this, LGTM

Don't get mad at me, but the tests are failing because I think there's just some complications in earlier versions (due to prolly me), but with the latest constraints from the validator and spec, the "correct" way of doing things is at least more clear. However, had to adjust a few more lines... :p 

Pushing up soon.

File: examples/convert_ieeg_to_bids.py [hunk 12]
```diff
@@ -83,10 +83,12 @@
     elec[:, ind] = list(map(float, electrode_tsv[axis]))
 
 ###############################################################################
-# Now we make a montage stating that the iEEG contacts are in MRI
-# coordinate system.
+# Now we make a montage stating that the iEEG contacts are in MNI coordinate
+# system, which corresponds to the `fsaverage` subject in FreeSurfer.
+# For example, one can look at how mne-python deals with iEEG data at
+# `Working with SEEG <https://mne.tools/stable/auto_tutorials/misc/plot_seeg.html>`_  # noqa
```
Comment (hoechenberger) [inline examples/convert_ieeg_to_bids.py:89]:
```suggestion
# Now we make a montage stating that the iEEG contacts are in the MNI 
# coordinate system, which corresponds to the `fsaverage` subject in
# FreeSurfer. For example, one can look at how mne-python deals with iEEG data
# at `Working with SEEG 
# <https://mne.tools/stable/auto_tutorials/misc/plot_seeg.html>`_
```

I think the line break should work with the link

File: mne_bids/config.py [hunk 13]
```diff
@@ -161,32 +161,42 @@
...
+    'ACPC': 'ras',  # XXX: there is no ACPC in mne-python and so this is a one-way mapping from BIDS -> MNE  # noqa
+    'fsaverage': 'mni_tal',  # XXX: note fsaverage and MNI305 are the same  # noqa
+    'MNI305': 'mni_tal',
+    'Other': 'unknown'
+}
+
+# mapping from supported MNE coordinate frames -> BIDS
+# XXX: note that there are a lot fewer MNE available coordinate
+# systems so the range of BIDS supported coordinate systems we
+# can write is limited.
```
Comment (hoechenberger) [inline mne_bids/config.py:190]:
```suggestion
# XXX: note that there are a lot fewer MNE coordinate
# systems, so the range of BIDS-supported coordinate systems we
# can write is limited.
```

File: mne_bids/config.py [hunk 14]
```diff
@@ -221,18 +231,9 @@
...
-                 'is at the center of a 256×256×256 voxel anisotropic '
-                 'volume (may not be in the center of the head).',
-    'mni_tal': 'MNI template in Talairach coordinates',
-    'fs_tal': 'Freesurfer template in Talairach coordinates',
-    'ras': 'RAS means that the first dimension (X) points towards '
-           'the right hand side of the head, the second dimension (Y) '
-           'points towards the Anterior aspect of the head, and the '
-           'third dimension (Z) points towards the top of the head.',
+    'fsaverage': 'Defined by Freesurfer, the MRI (surface RAS) origin is '
+                 'at the center of a 256×256×256 1mm anisotropic volume '
```
Comment (hoechenberger) [inline mne_bids/config.py:235]:
```suggestion
    'fsaverage': 'Defined by FreeSurfer, the MRI (surface RAS) origin is '
                 'at the center of a 256×256×256 mm^3 anisotropic volume '
```

File: mne_bids/dig.py [hunk 15]
```diff
@@ -396,19 +365,18 @@ def _write_dig_bids(bids_path, raw, overwrite=False, verbose=True):
     if datatype == "ieeg":
```
Comment (hoechenberger) [inline mne_bids/dig.py:365]:
```suggestion
    if datatype == 'ieeg':
```

File: mne_bids/dig.py [hunk 16]
```diff
@@ -451,86 +419,103 @@ def _read_dig_bids(electrodes_fpath, coordsystem_fpath,
...
-            coord_frame = BIDS_TO_MNE_FRAMES.get(coord_frame, None)
+            coord_frame = BIDS_TO_MNE_FRAMES.get(bids_coord_frame, None)
     elif datatype == 'ieeg':
-        if coord_frame not in BIDS_IEEG_COORDINATE_FRAMES:
-            warn("iEEG Coordinate frame is not accepted "
-                 "BIDS keyword. The allowed keywords are: "
-                 "{}".format(BIDS_IEEG_COORDINATE_FRAMES))
-            coord_frame = None
-        elif coord_frame == 'pixels':
+        # ieeg datatype for BIDS only supports
```
Comment (hoechenberger) [inline mne_bids/dig.py:454]:
```suggestion
        # iEEG datatype for BIDS only supports
```

File: mne_bids/dig.py [hunk 17]
```diff
@@ -451,86 +419,103 @@ def _read_dig_bids(electrodes_fpath, coordsystem_fpath,
...
+        # acpc, pixels and then standard templates
+        # iEEG datatype for mne-python only supports
+        # mni_tal == fsaverage / MNI305
```
Comment (hoechenberger) [inline mne_bids/dig.py:457]:
```suggestion
        # mni_tal == fsaverage == MNI305
```
reduce potential for confusion ;)

File: mne_bids/dig.py [hunk 18]
```diff
@@ -451,86 +419,103 @@ def _read_dig_bids(electrodes_fpath, coordsystem_fpath,
...
+    if montage is not None:
+        # determine if there are problematic channels
+        ch_pos = montage._get_ch_pos()
+        nan_chs = []
+        for ch_name, ch_coord in ch_pos.items():
+            if any(np.isnan(ch_coord)) and ch_name not in raw.info['bads']:
+                nan_chs.append(ch_name)
+        if len(nan_chs) > 0:
+            warn("There are channels without locations "
+                 "(n/a) that are not marked as bad: {}".format(nan_chs))
```
Comment (hoechenberger) [inline mne_bids/dig.py:516]: Can you turn this into an f-string, please?

File: mne_bids/tests/test_write.py [hunk 19]
```diff
@@ -2187,8 +2187,8 @@ def test_event_storage():
 @pytest.mark.parametrize(
     'dir_name, fname, reader, datatype, coord_frame', [
         ('EDF', 'test_reduced.edf', _read_raw_edf, 'ieeg', 'mni_tal'),
-        ('EDF', 'test_reduced.edf', _read_raw_edf, 'ieeg', 'fs_tal'),
-        ('EDF', 'test_reduced.edf', _read_raw_edf, 'ieeg', 'unknown'),
+        # ('EDF', 'test_reduced.edf', _read_raw_edf, 'ieeg', 'fs_tal'),  # XXX: shouldn't be used  # noqa
```
Comment (hoechenberger) [inline mne_bids/tests/test_write.py:2190]: Why not?

Comment (adam2392) [reply mne_bids/tests/test_write.py:2190]: Because of https://github.com/mne-tools/mne-bids/pull/704#discussion_r583889948

But I can add this back in I suppose and just have this set to "other"?

Comment (hoechenberger) [reply mne_bids/tests/test_write.py:2190]: Oh the "shouldn't be used" is referring to `fs_tal`, not to the actual test itself.

Well… why don't you simply remove this entire line, then?

**Review by hoechenberger — APPROVED** [review]
Aside from this one comment, LGTM

**Review by hoechenberger — APPROVED** [review]

**Comment by hoechenberger** [comment]
Thanks, @adam2392!
